qiskit/transpiler/passes/basis/unroller.py

```diff
@@ -68,6 +70,22 @@ def run(self, dag):
                 else:
                     continue
 
+            if isinstance(node.op, ControlFlowOp):
+                new_params = []
+                for param in node.op.params:
+                    if isinstance(param, QuantumCircuit):
+                        # TODO: check whether unrolling is necessary
+                        dag_block = circuit_to_dag(param)
+                        unrolled_dag_block = self.run(dag_block)
+                        unrolled_circ_block = dag_to_circuit(unrolled_dag_block)
+                        new_params.append(unrolled_circ_block)
+                    else:
+                        new_params.append(copy.copy(param))
+                new_cf_op = node.op.copy_no_body()
+                new_cf_op.params = new_params
+                node.op = new_cf_op
```

Comment:
This construction feels a little awkward - all the circuit-specific blocks are immediately available as `ControlFlowOp.blocks`, and it feels odd to have a constructor routine that requires mutation immediately afterwards.  I think it might be neater to exchange the `ControlFlowOp.copy_no_body` method with something like
```python
class ForLoopOp:
    def replace_blocks(self, blocks):
        (body,) = blocks
        is not isinstance(body, QuantumCircuit):
            ...
        return ForLoopOp(self._params[0], self._params[1], body)
```
(and similar for the other ops.  It'd need to be added as an abstract method on `ControlFlowOp` too)

This avoids constructing an unnecessary `QuantumCircuit`, and it lets the class itself be responsible for managing the parameter list, rather than needing to manually do it.  For example, `ForLoopOp` knows that none of its parameters need copying (`Parameter` is immutable, and the indexset is either a range or tuple, which are immutable).

The unroller loop would then be much simpler.  Something like:
```python
if isinstance(node.op, ControlFlowOp):
    node.op = node.op.replace_blocks(
        dag_to_circuit(self.run(circuit_to_dag(block)))
        for block in node.op.blocks
    )
```

Reply:
Yes, I was attempting to avoid using a block setter in the spirit of immutability as we discussed but really the params setter is the same but less convenient.